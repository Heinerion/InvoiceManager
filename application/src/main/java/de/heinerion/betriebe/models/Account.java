package de.heinerion.betriebe.models;

import java.util.Objects;

public final class Account {
  private String bic;
  private String iban;
  private String name;

  /**
   * For persistence only
   */
  public Account() {
  }

  public Account(String aName, String aBic, String anIban) {
    this.name = aName;
    this.bic = aBic;
    this.iban = anIban;
  }

  public String getBic() {
    return bic;
  }

  public void setBic(String bic) {
    this.bic = bic;
  }

  public String getIban() {
    return iban;
  }

  public void setIban(String iban) {
    this.iban = iban;
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Account account = (Account) o;
    return Objects.equals(bic, account.bic) &&
        Objects.equals(iban, account.iban) &&
        Objects.equals(name, account.name);
  }

  @Override
  public int hashCode() {
    return Objects.hash(bic, iban, name);
  }

  @Override
  public String toString() {
    return name;
  }
}
