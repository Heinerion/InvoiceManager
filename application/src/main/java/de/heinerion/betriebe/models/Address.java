package de.heinerion.betriebe.models;

import java.util.Objects;
import java.util.Optional;

public final class Address {
  private String apartment;
  private String company;
  private String district;
  private String location;
  private String number;
  private String postalCode;
  private String recipient;
  private String street;

  public Optional<String> getApartment() {
    return Optional.ofNullable(apartment);
  }

  public Optional<String> getCompany() {
    return Optional.ofNullable(company);
  }

  public Optional<String> getDistrict() {
    return Optional.ofNullable(district);
  }

  public String getLocation() {
    return this.location;
  }

  public String getNumber() {
    return this.number;
  }

  public String getPostalCode() {
    return this.postalCode;
  }

  public String getRecipient() {
    return this.recipient;
  }

  public String getStreet() {
    return this.street;
  }

  public void setApartment(String apartment) {
    if (isValidMessage(apartment)) {
      this.apartment = apartment;
    }
  }

  public void setCompany(String company) {
    if (isValidMessage(company)) {
      this.company = company;
    }
  }

  public void setDistrict(String district) {
    if (isValidMessage(district)) {
      this.district = district;
    }
  }

  public void setLocation(String location) {
    this.location = location;
  }

  public void setNumber(String number) {
    this.number = number;
  }

  public void setPostalCode(String postalCode) {
    this.postalCode = postalCode;
  }

  public void setRecipient(String recipient) {
    this.recipient = recipient;
  }

  public void setStreet(String street) {
    this.street = street;
  }

  private boolean isValidMessage(String message) {
    return null != message && !message.trim().isEmpty();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Address address = (Address) o;
    return Objects.equals(apartment, address.apartment) &&
        Objects.equals(company, address.company) &&
        Objects.equals(district, address.district) &&
        Objects.equals(location, address.location) &&
        Objects.equals(number, address.number) &&
        Objects.equals(postalCode, address.postalCode) &&
        Objects.equals(recipient, address.recipient) &&
        Objects.equals(street, address.street);
  }

  @Override
  public int hashCode() {
    return Objects.hash(apartment, company, district, location, number, postalCode, recipient, street);
  }

  @Override
  public String toString() {
    return recipient;
  }
}
