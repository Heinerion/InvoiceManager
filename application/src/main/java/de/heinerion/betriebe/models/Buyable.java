package de.heinerion.betriebe.models;

interface Buyable {
  String getName();

  double getPricePerUnit();

  String getUnit();
}
