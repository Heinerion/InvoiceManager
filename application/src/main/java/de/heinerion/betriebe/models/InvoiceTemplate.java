package de.heinerion.betriebe.models;

import java.io.Serial;
import java.io.Serializable;
import java.text.Collator;
import java.util.Arrays;
import java.util.UUID;

public class InvoiceTemplate implements Serializable, Comparable<InvoiceTemplate> {
  @Serial
  private static final long serialVersionUID = 5654884407643922708L;

  private String name = "";
  private String[][] inhalt;
  private UUID companyId;

  @Override
  public final int compareTo(InvoiceTemplate o) {
    return (Collator.getInstance()).compare(this.getName(), o.getName());
  }

  @Override
  public boolean equals(Object obj) {
    return obj instanceof InvoiceTemplate other
        && compareTo(other) == 0;
  }

  @Override
  public int hashCode() {
    return getName().hashCode();
  }

  public UUID getCompanyId() {
    return companyId;
  }

  public void setCompanyId(UUID companyId) {
    this.companyId = companyId;
  }

  public final String[][] getInhalt() {
    return inhalt;
  }

  public final String getName() {
    return this.name;
  }

  public final void setInhalt(String[][] content) {
    this.inhalt = Arrays.copyOf(content, content.length);
  }

  public final void setName(String aName) {
    this.name = aName;
  }

  @Override
  public final String toString() {
    return this.getName();
  }
}
