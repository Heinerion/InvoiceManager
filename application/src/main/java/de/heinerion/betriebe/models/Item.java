package de.heinerion.betriebe.models;

import java.util.Objects;

public final class Item implements Buyable {
  private String name;
  private String unit;
  private double pricePerUnit;
  private double quantity;
  private double total;

  /**
   * For persistence only
   */
  public Item() {
  }

  public Item(String name, String unit, double pricePerUnit) {
    this(name, unit, pricePerUnit, 0);
  }

  public Item(String name, String unit, double pricePerUnit, double aQuantity) {
    this.name = name;
    this.unit = unit;
    this.pricePerUnit = pricePerUnit;
    this.quantity = aQuantity;
    this.updateValues();
  }

  public void addQuantity(double additionalQuantity) {
    this.quantity += additionalQuantity;
    this.updateValues();
  }

  @Override
  public String getName() {
    return name;
  }

  public void setName(String newName) {
    this.name = newName;
  }

  @Override
  public double getPricePerUnit() {
    return pricePerUnit;
  }

  public void setPricePerUnit(double newPrice) {
    this.pricePerUnit = newPrice;
  }

  public double getQuantity() {
    return this.quantity;
  }

  public void setQuantity(double newQuantity) {
    this.quantity = newQuantity;
    this.updateValues();
  }

  public double getTotal() {
    return this.total;
  }

  public void setTotal(double total) {
    this.total = total;
  }

  @Override
  public String getUnit() {
    return unit;
  }

  public void setUnit(String newUnit) {
    unit = newUnit;
  }

  @Override
  public String toString() {
    return this.getName();
  }

  private void updateValues() {
    final double pricePerUnit = this.getPricePerUnit();
    this.total = pricePerUnit * this.quantity;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Item item = (Item) o;
    return Double.compare(item.quantity, quantity) == 0 &&
        Double.compare(item.total, total) == 0 &&
        Objects.equals(name, item.name) &&
        Objects.equals(pricePerUnit, item.pricePerUnit) &&
        Objects.equals(unit, item.unit);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, pricePerUnit, unit, quantity, total);
  }
}
