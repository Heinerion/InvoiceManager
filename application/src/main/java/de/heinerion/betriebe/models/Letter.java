package de.heinerion.betriebe.models;

import de.heinerion.invoice.storage.PathTools;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Letter implements Storable {
  private final Company company;
  private final LocalDate date;

  String subject;

  private final List<String> messageLines;
  final Address receiver;

  public Letter(LocalDate date, Company sender, Address receiver) {
    this.date = date;
    this.company = sender;
    this.receiver = receiver;

    this.messageLines = new ArrayList<>();
  }

  public void addMessageLine(String messageLine) {
    this.messageLines.add(messageLine);
  }

  @Override
  public String[] getClassification() {
    return new String[]{this.company.getDescriptiveName(),
        PathTools.determineFolderName(this.getClass())};
  }

  public Company getCompany() {
    return this.company;
  }

  public LocalDate getDate() {
    return this.date;
  }

  @Override
  public String getEntryName() {
    return this.getDate().toString() + " - "
        + this.getReceiver().getRecipient();
  }

  public List<String> getMessageLines() {
    return this.messageLines;
  }

  public Address getReceiver() {
    return this.receiver;
  }

  public String getSubject() {
    return this.subject;
  }

  public boolean isPrintable() {
    final boolean hasSubject = !(subject == null || subject.trim().isEmpty());
    final boolean hasContent = !messageLines.isEmpty();
    return hasSubject && hasContent;
  }

  public void setSubject(String subject) {
    this.subject = subject;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Letter letter = (Letter) o;
    return Objects.equals(company, letter.company) &&
        Objects.equals(date, letter.date) &&
        Objects.equals(subject, letter.subject) &&
        Objects.equals(messageLines, letter.messageLines) &&
        Objects.equals(receiver, letter.receiver);
  }

  @Override
  public int hashCode() {
    return Objects.hash(company, date, subject, messageLines, receiver);
  }

  @Override
  public String toString() {
    return "Letter{" +
        "company=" + company +
        ", date=" + date +
        ", subject='" + subject + '\'' +
        ", messageLines=" + messageLines +
        ", receiver=" + receiver +
        '}';
  }
}
