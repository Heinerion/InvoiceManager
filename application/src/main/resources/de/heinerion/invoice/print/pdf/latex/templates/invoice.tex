<#assign invoiceItems=invoice.items?size >
<#assign invoiceItemsMax=5 >
<#assign itemsWithUnit = [] />
<#list invoice.items as item>
    <#if item.unit??>
        <#assign itemsWithUnit = itemsWithUnit + [item.name] />
    </#if>
</#list>
\documentclass[fontsize=12pt, fromalign=center, fromphone=true, paper=a4]{scrlttr2}
\usepackage[utf8]{inputenc}
\usepackage[ngermanb]{babel}
\usepackage[right]{eurosym}
\usepackage{longtable}
\usepackage{lastpage}             % determine last page
\usepackage{scrlayer-scrpage}     % scr-Styling
\usepackage[hidelinks]{hyperref}  % use hyperref without link highlighting
\hypersetup{
  pdftitle={${labels.title}},
  pdfauthor={${invoice.company.officialName}},
  pdfsubject={<#list itemsWithUnit as item>${item}<#sep>, </#list>},
  <#setting locale="en_US">
  pdfkeywords={${invoice.gross}}
  <#setting locale="de_DE">
}
\renewcommand{\raggedsignature}{\raggedright}
\setkomavar{signature}{\underline{${labels.signature}:\hspace{10cm}}}
\setkomavar{subject}{${invoice.subject}}
\setkomavar{fromaddress}{${invoice.company.address.street} ${invoice.company.address.number}, ${invoice.company.address.postalCode} ${invoice.company.address.location}}
\setkomavar{fromphone}{${invoice.company.phoneNumber}}
\setkomavar{fromname}{{${invoice.company.officialName}}\tiny}
\date{}
\setkomavar{location}{
  \vfill
  \raggedright
  \footnotesize
  \begin{tabular}{ll}
  \textsc ${labels.date} & : ${invoice.date?date("yyyy-MM-dd")}\\
  \textsc ${labels.number} & : ${invoice.number?string["0"]}\\
  \textsc ${labels.taxNumber} & : ${invoice.company.taxNumber}\\
  \textsc BIC & : ${invoice.company.account.bic}\\
  \textsc IBAN & : ${invoice.company.account.iban}\\
  \multicolumn{2}{l}{${invoice.company.account.name}}
  \end{tabular}
}

\setkomavar{firstfoot}{
  \usekomafont{pageheadfoot}
  \parbox{\useplength{firstfootwidth}}{
    \rule{\linewidth}{.4pt}\\
    \parbox[t]{0.8\linewidth}{
      ${invoice.company.officialName}, ${labels.number}: ${invoice.number?string["0"]}, ${invoice.date?date("yyyy-MM-dd")}
    }
    \hfill
    \parbox[t]{0.1\linewidth}{
      Seite \thepage \hspace{1pt} von \pageref{LastPage}
    }
  }
}
\setkomafont{pageheadfoot}{\scriptsize}

% Use the footer-style of the first page in following pages
\DeclareNewLayer[
  foreground,
  textarea,
  voffset=\useplength{firstfootvpos},
  hoffset=\dimexpr.5\paperwidth-.5\useplength{firstfootwidth}\relax,
  width=\useplength{firstfootwidth},
  mode=picture,
  contents=\putUL{\raisebox{\dimexpr-\height}{\usekomavar{firstfoot}}}
]{likefirstpage.foot}

\AddLayersToPageStyle{scrheadings}{likefirstpage.foot}
\clearpairofpagestyles

\usepackage{etoolbox}% http://ctan.org/pkg/etoolbox
\makeatletter

% allows for more horizontal space in the location
\@setplength{lochpos}{\oddsidemargin}
\@addtoplength{lochpos}{6cm}

% pull signature closer to the table
\patchcmd{\closing}% <cmd>
  {\parbox}% <search>
  {\parbox{\linewidth}{\raggedsignature\strut\ignorespaces\let\\\relax%
      #1 \usekomavar{signature}}%
   \@gobbletwo}%< <replace>
  {}{}% <success><failure>
\makeatother

\begin{document}
\begin{letter}{<#list address as line>${line}<#sep>\\
</#list>
}

\opening{}\vspace{-25pt}
\begin{longtable}{p{5cm}p{1cm}p{3cm}|r|r|}
\hline
\multicolumn{3}{|c|}{\textsc ${labels.description}}&\multicolumn{1}{c|}{\textsc ${labels.pricePerUnit}}&\multicolumn{1}{c|}{\textsc ${labels.sum}}\\
\hline
<#list invoice.items as item>
\hline
    <#if item.unit??>
\multicolumn{1}{|l}{${item.name}}&\multicolumn{1}{r}{${item.quantity}}&${item.unit}&\hspace{12pt} รก \EUR{${item.pricePerUnit}}&\EUR{${item.total}}\\
    <#else>
\multicolumn{2}{|l}{${item.name}}&&&\\
    </#if>
</#list>
<#if invoiceItems < invoiceItemsMax>
  <#list invoiceItems..invoiceItemsMax as number>
\hline
    <#if number < invoiceItemsMax>
\multicolumn{1}{|l}{$\phantom{sth}$}&&&&\\
    <#else>
\multicolumn{1}{|l}{$\phantom{sth}$}&\multicolumn{1}{r}{\phantom{(\hfill000,00}}&&&\\
    </#if>
  </#list>
<#else>
\hline
\multicolumn{1}{|l}{$\phantom{sth}$}&\multicolumn{1}{r}{\phantom{(\hfill000,00}}&&&\\
</#if>
\hline
\multicolumn{1}{|l}{${labels.net}}&&&&\EUR{${invoice.net}}\\
\hline
&&&\multicolumn{1}{|c|}{${invoice.vat}\% ${labels.vat}}&\EUR{${invoice.tax}}\\\cline{4-5}
&&&\multicolumn{1}{|l|}{\textsc ${labels.sum}}&\EUR{${invoice.gross}}\\\cline{4-5}
\end{longtable}
\vfill
\closing{}

\end{letter}
\end{document}