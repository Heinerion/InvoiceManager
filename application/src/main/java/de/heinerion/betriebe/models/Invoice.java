package de.heinerion.betriebe.models;

import de.heinerion.invoice.Translator;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Invoice extends Letter {
  private static final int PERCENT = 100;

  private int number;

  private List<Item> items;

  private double net;
  private double vat;
  private double tax;
  private double gross;

  /**
   * For persistence only
   */
  public Invoice() {
    super();
  }

  public Invoice(LocalDate aDate, Company theSender, Address theReceiver) {
    super(aDate, theSender, theReceiver);
    subject = Translator.translate("invoice.title");

    this.items = new ArrayList<>();

    this.vat = theSender.getValueAddedTax();
    this.updateValues();

    this.number = theSender.getInvoiceNumber();
  }

  public void add(String article, String unit, double price, double count) {
    this.addItem(new Item(article, unit, price, count));
  }

  @Override
  public void addMessageLine(String messageLine) {
    add(messageLine, null, 0, 0);
  }

  private void addItem(Item item) {
    this.items.add(item);
    this.updateValues();
  }

  // TODO removeItem?

  @Override
  public String getEntryName() {
    return this.number + " " + this.receiver.getRecipient();
  }

  public double getGross() {
    return this.gross;
  }

  public void setGross(double gross) {
    this.gross = gross;
  }

  public List<Item> getItems() {
    return this.items;
  }

  public void setItems(List<Item> items) {
    this.items = items;
  }

  public double getNet() {
    return this.net;
  }

  public void setNet(double net) {
    this.net = net;
  }

  public int getNumber() {
    return this.number;
  }

  public void setNumber(int number) {
    this.number = number;
  }

  public double getTax() {
    return this.tax;
  }

  public void setTax(double tax) {
    this.tax = tax;
  }

  public double getVat() {
    return this.vat;
  }

  public void setVat(double vat) {
    this.vat = vat;
  }

  @Override
  public boolean isPrintable() {
    return true;
  }

  private void updateValues() {
    this.net = 0;
    for (Item item : this.getItems()) {
      this.net += item.getTotal();
    }
    this.tax = this.net * this.vat / PERCENT;
    this.gross = this.net + this.tax;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Invoice invoice = (Invoice) o;
    return number == invoice.number &&
        Double.compare(invoice.net, net) == 0 &&
        Double.compare(invoice.vat, vat) == 0 &&
        Double.compare(invoice.tax, tax) == 0 &&
        Double.compare(invoice.gross, gross) == 0 &&
        Objects.equals(items, invoice.items);
  }

  @Override
  public int hashCode() {
    return Objects.hash(number, items, net, vat, tax, gross);
  }

  @Override
  public String toString() {
    return "Invoice{" +
        "number=" + number +
        ", items=" + items +
        ", net=" + net +
        ", vat=" + vat +
        ", tax=" + tax +
        ", gross=" + gross +
        ", subject='" + subject + '\'' +
        ", receiver=" + receiver +
        '}';
  }
}
