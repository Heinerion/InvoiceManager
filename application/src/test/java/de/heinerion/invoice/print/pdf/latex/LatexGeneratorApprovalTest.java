package de.heinerion.invoice.print.pdf.latex;

import de.heinerion.invoice.models.*;
import de.heinerion.invoice.testsupport.builder.*;
import org.approvaltests.Approvals;
import org.junit.*;

import java.time.*;

public class LatexGeneratorApprovalTest {
  private static final Month MONTH = Month.JUNE;
  private static final int YEAR = 2010;
  private static final int DAY = 25;

  private static Company sender;
  private static Address receiverAddress;
  private static LocalDate date;

  private static Letter letter;
  private static Invoice invoice;

  @BeforeClass
  public static void prepare() {
    sender = new CompanyBuilder()
        .withAddress(new AddressBuilder()
            .withName("Bob's Address")
            .withBlock("Bob Bubbles\n\nstreet number\npostalCode location"))
        .build();
    receiverAddress = new AddressBuilder()
        .withName("Receiver's Address")
        .withBlock("Receiver\nStreet 7\n\nPC Somewhere©")
        .build();
    date = LocalDate.of(YEAR, MONTH, DAY);
  }

  @Before
  public final void setUp() {
    letter = new Letter(date, sender, receiverAddress);
    letter.setSubject("Test");

    invoice = new Invoice(date, sender, receiverAddress, sender.getInvoiceNumber());
    invoice.addItem(Item.of(invoice.getItems().size(), "Artikel 1", "Stück", 1.50, 2));
  }

  protected LatexGenerator getLatexGenerator() {
    return new LatexGeneratorFreeMarkerImpl();
  }

  @Test
  public final void generateEmptyLetter() {
    final String result = getLatexGenerator().generateSourceContent(letter);

    Approvals.verify(result);
  }

  @Test
  public final void generateNonEmptyLetter() {
    String content = "Hello new World";
    letter.addMessageLine(content);
    letter.addMessageLine(content);
    letter.addMessageLine(content);
    letter.addMessageLine(content);
    String result = getLatexGenerator().generateSourceContent(letter);

    Approvals.verify(result);
  }

  @Test
  public final void generateInvoice() {
    final String result = getLatexGenerator().generateSourceContent(invoice);

    Approvals.verify(result);
  }

  @Test
  public final void generateInvoice_items() {
    invoice.addItem(Item.of(invoice.getItems().size(), "Artikel 2", "Stück", 3.44, 3));
    final String result = getLatexGenerator().generateSourceContent(invoice);

    Approvals.verify(result);
  }

  @Test
  public final void generateInvoice_text() {
    invoice.addItem(Item.of(invoice.getItems().size(), "Message"));

    final String result = getLatexGenerator().generateSourceContent(invoice);

    Approvals.verify(result);
  }
}
