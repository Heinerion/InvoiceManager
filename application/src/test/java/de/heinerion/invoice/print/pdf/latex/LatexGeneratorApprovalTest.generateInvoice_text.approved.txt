\documentclass[fontsize=12pt, fromalign=center, fromphone=true, paper=a4]{scrlttr2}
\usepackage[utf8]{inputenc}
\usepackage[ngermanb]{babel}
\usepackage[right]{eurosym}
\usepackage{longtable}
\usepackage{lastpage}             % determine last page
\usepackage{scrlayer-scrpage}     % scr-Styling
\usepackage[hidelinks]{hyperref}  % use hyperref without link highlighting
\hypersetup{
  pdftitle={Rechnung},
  pdfauthor={officialName},
  pdfsubject={Artikel 1},
  pdfkeywords={3.30}
}
\renewcommand{\raggedsignature}{\raggedright}
\setkomavar{signature}{\underline{Unterschrift:\hspace{10cm}}}
\setkomavar{subject}{Rechnung}
\setkomavar{fromaddress}{street number, postalCode location}
\setkomavar{fromphone}{123-456}
\setkomavar{fromname}{{officialName}\tiny}
\date{}
\setkomavar{location}{
  \vfill
  \raggedright
  \footnotesize
  \begin{tabular}{ll}
  \textsc Datum & : 25.06.2010\\
  \textsc Rechnungs-Nr. & : 0\\
  \textsc Steuernummer & : taxNumber\\
  \textsc BIC & : bic\\
  \textsc IBAN & : iban\\
  \multicolumn{2}{l}{institute}
  \end{tabular}
}

\setkomavar{firstfoot}{
  \usekomafont{pageheadfoot}
  \parbox{\useplength{firstfootwidth}}{
    \rule{\linewidth}{.4pt}\\
    \parbox[t]{0.8\linewidth}{
      officialName, Rechnungs-Nr.: 0, 25.06.2010
    }
    \hfill
    \parbox[t]{0.1\linewidth}{
      Seite \thepage \hspace{1pt} von \pageref{LastPage}
    }
  }
}
\setkomafont{pageheadfoot}{\scriptsize}

% Use the footer-style of the first page in following pages
\DeclareNewLayer[
  foreground,
  textarea,
  voffset=\useplength{firstfootvpos},
  hoffset=\dimexpr.5\paperwidth-.5\useplength{firstfootwidth}\relax,
  width=\useplength{firstfootwidth},
  mode=picture,
  contents=\putUL{\raisebox{\dimexpr-\height}{\usekomavar{firstfoot}}}
]{likefirstpage.foot}

\AddLayersToPageStyle{scrheadings}{likefirstpage.foot}
\clearpairofpagestyles

\usepackage{etoolbox}% http://ctan.org/pkg/etoolbox
\makeatletter

% allows for more horizontal space in the location
\@setplength{lochpos}{\oddsidemargin}
\@addtoplength{lochpos}{6cm}

% pull signature closer to the table
\patchcmd{\closing}% <cmd>
  {\parbox}% <search>
  {\parbox{\linewidth}{\raggedsignature\strut\ignorespaces\let\\\relax%
      #1 \usekomavar{signature}}%
   \@gobbletwo}%< <replace>
  {}{}% <success><failure>
\makeatother

\begin{document}
\begin{letter}{Receiver\\
Street 7\\
PC Somewhere©}

\opening{}\vspace{-25pt}
\begin{longtable}{p{5cm}p{1cm}p{3cm}|r|r|}
\hline
\multicolumn{3}{|c|}{\textsc Bezeichnung}&\multicolumn{1}{c|}{\textsc Einzelpreis}&\multicolumn{1}{c|}{\textsc Gesamt}\\
\hline
\hline
\multicolumn{1}{|l}{Artikel 1}&\multicolumn{1}{r}{2,00}&Stück&\hspace{12pt} á \EUR{1,50}&\EUR{3,00}\\
\hline
\multicolumn{2}{|l}{Message}&&&\\
\hline
\multicolumn{1}{|l}{$\phantom{sth}$}&&&&\\
\hline
\multicolumn{1}{|l}{$\phantom{sth}$}&&&&\\
\hline
\multicolumn{1}{|l}{$\phantom{sth}$}&&&&\\
\hline
\multicolumn{1}{|l}{$\phantom{sth}$}&\multicolumn{1}{r}{\phantom{(\hfill000,00}}&&&\\
\hline
\multicolumn{1}{|l}{Netto}&&&&\EUR{3,00}\\
\hline
&&&\multicolumn{1}{|c|}{10,00\% MwSt}&\EUR{0,30}\\\cline{4-5}
&&&\multicolumn{1}{|l|}{\textsc Gesamt}&\EUR{3,30}\\\cline{4-5}
\end{longtable}
\vfill
\closing{}

\end{letter}
\end{document}